pulp_container/app/viewsets.py

```diff
@@ -462,6 +462,54 @@ def untag(self, request, pk):
         return OperationPostponedResponse(result, request)
 
 
+class SignOperationsMixin:
+    """Signing mixing for both types of the repos."""
+
+    @extend_schema(
+        description="Trigger an asynchronous task to sign content.",
+        summary="Sign images in the repo",
+        responses={202: AsyncOperationResponseSerializer},
+    )
+    @action(detail=True, methods=["post"], serializer_class=serializers.RepositorySignSerializer)
+    def sign(self, request, pk):
+        """
+        Signs manifests by tag in the repository.
+        """
+        repository = self.get_object()
+        serializer = serializers.RepositorySignSerializer(
+            data=request.data, context={"request": request, "repository_pk": pk}
+        )
+
+        # Validate synchronously to return 400 errors.
+        serializer.is_valid(raise_exception=True)
+        signing_service = serializer.validated_data.get(
+            "manifest_signing_service", repository.manifest_signing_service
+        )
+        reference = serializer.validated_data.get("reference")
+
+        tags_list = request.data.get("tags_list")
+        if tags_list:
+            tags_list_pks = models.Tag.objects.filter(name__in=tags_list).values_list(
+                "pk", flat=True
+            )
+            tags_list_pks = list(tags_list_pks)
+        else:
+            tags_list_pks = None
+        result = dispatch(
+            tasks.sign,
+            # FIXME LookupError: Could not determine ViewSet base name for model
+            # shared_resources=[signing_service],
+            exclusive_resources=[repository],
+            kwargs={
+                "repository_pk": repository.pk,
+                "reference": reference,
+                "signing_service_pk": signing_service.pk,
+                "tags_list": tags_list_pks,
```

Comment:
@rochacbruno @newswangerd hey, here a draft PR which is quite close to be ready for  review. It need tests. And docs, including the example of the bash script for the signing service.
Please take a look. Current approach is to whether sign whole repo, or  if specified list of tags, then sign the manifests that the tags are referencing. I think tags are more user friendly rather then asking to specify the manifest digest.

if you want to test it out then:
1. push an image to pulp with podman
2.  add manifest signing service
3. hit the repo `sign` endpoint

```$ podman push pulp3-source-fedora34.fluffy.example.com/test/minio
Getting image source signatures
Copying blob 1323ffbff4dd done  
Copying blob 188aada3e32b done  
Copying blob 6fbc01e60a74 done  
Copying blob 2b6e25df12be done  
Copying blob 744c86b54390 done  
Copying blob df0da73eeb7b done  
Copying blob 7eea1030eff3 done  
Copying config 01431999f9 done  
Writing manifest to image destination
Storing signatures
(pulp) [vagrant@pulp3-source-fedora34 pulp_container]$ pulpcore-manager add-signing-service test-service7 /home/vagrant/devel/pulp_container/bash-script.sh 45ACE14E3EBB9BBA --class container:ManifestSigningService
Successfully added signing service test-service7 for key 8F336A705074623F7FC3273945ACE14E3EBB9BBA.
(pulp) [vagrant@pulp3-source-fedora34 pulp_container]$ http GET https://pulp3-source-fedora34.fluffy.example.com/pulp/api/v3/repositories/container/container-push/
HTTP/1.1 200 OK
Access-Control-Expose-Headers: Correlation-ID
Allow: GET, HEAD, OPTIONS
Connection: keep-alive
Content-Length: 557
Content-Type: application/json
Correlation-ID: 435ddf57fb7748f98e09fa98c18655fb
Date: Fri, 10 Dec 2021 20:39:21 GMT
Referrer-Policy: same-origin
Server: nginx/1.20.1
Strict-Transport-Security: max-age=15768000
Vary: Accept, Cookie
X-Content-Type-Options: nosniff
X-Frame-Options: DENY

{
    "count": 1,
    "next": null,
    "previous": null,
    "results": [
        {
            "description": null,
            "latest_version_href": "/pulp/api/v3/repositories/container/container-push/3b279a32-b313-44bf-ad41-4359a92cae24/versions/9/",
            "manifest_signing_service": null,
            "name": "test/minio",
            "pulp_created": "2021-12-10T20:38:40.379570Z",
            "pulp_href": "/pulp/api/v3/repositories/container/container-push/3b279a32-b313-44bf-ad41-4359a92cae24/",
            "pulp_labels": {},
            "retain_repo_versions": null,
            "versions_href": "/pulp/api/v3/repositories/container/container-push/3b279a32-b313-44bf-ad41-4359a92cae24/versions/"
        }
    ]
}


(pulp) [vagrant@pulp3-source-fedora34 pulp_container]$ http GET https://pulp3-source-fedora34.fluffy.example.com/pulp/api/v3/signing-services/
HTTP/1.1 200 OK
Access-Control-Expose-Headers: Correlation-ID
Allow: GET, HEAD, OPTIONS
Connection: keep-alive
Content-Length: 2829
Content-Type: application/json
Correlation-ID: 380d13ca59f04211a0af14267e84f316
Date: Fri, 10 Dec 2021 20:39:29 GMT
Referrer-Policy: same-origin
Server: nginx/1.20.1
Strict-Transport-Security: max-age=15768000
Vary: Accept, Cookie
X-Content-Type-Options: nosniff
X-Frame-Options: DENY

{
    "count": 1,
    "next": null,
    "previous": null,
    "results": [
        {
            "name": "test-service7",
            "pubkey_fingerprint": "8F336A705074623F7FC3273945ACE14E3EBB9BBA",
            "public_key": "-----BEGIN PGP PUBLIC KEY BLOCK-----\n\nmQGNBGGerHwBDAC8AfkormXOinbco5gJz/H1KBu3F4/VEwsSynpix86K8ItbSnzk\nSrJj9GcuhvecNvVYC3tArma9uhiuZDQCXBBj7VBmZAR0sbbDZBrh93EnZgnNSSzM\n4bFJpuoYjUzghIkecuApl4KBtIXm7kM0tqzmgng+v0YttNb68RRwYokV3CRR6q23\nHxDnyD7DKc3jSEi36f59wg3TwkMIPbQHt3m79ARR1tT2KmCYEqOmnyIpWTszCBC3\nU9H+oRktskui/brDPdDOUp/2uOOZRxuhp2WFGV588GqrJs0p0OiucOjP2KlsNrr9\nC9t9wE3QqGj01g2TaiblDAaOzYijzfk/g0LqOf9tvZqcyX/F6QYtW7oIZH8xJSfP\nTsN6nXVbVfvoGtTRuRw1W1ij/nyC6wCVaFndB90jY0ywNE+l5uDvfCNv2icqq/rH\nfRDM3mRj2ZwLV/tRp0ZXe9ondmZbi25KE1nfkOMHIOplYtimO5QBtGFpleV4uPJw\nVOqaHJyHTp0KUz0AEQEAAbQaUHVwc2lrIDxwdXBzaWtAcmVkaGF0LmNvbT6JAdQE\nEwEIAD4WIQSPM2pwUHRiP3/DJzlFrOFOPrubugUCYZ6sfAIbAwUJA8JnAAULCQgH\nAgYVCgkICwIEFgIDAQIeAQIXgAAKCRBFrOFOPrubuiodC/sFy7/KnDeJ+bfsR4nc\nMBYXPtSU4cxbsyr5U++uOi6DHFUvvd7UFpdK3Cubs5G5MR/NYU2oCWeBQNTyzx5S\n9FhpNt5KTFhm56TgJApdcsC/rPbiR5/VRag5PxCji+UMLCAYKKbTrluTfFu2N6m1\no3B44eE3AaN2sCD3hf8L+ZFW1JYqe5w2/AS8B2/TjvWKeKBMwWWRJlHzpbzSfdjO\nuvc7UOrhjLQSU4a7T6VuUXoSs30IYAyIXrepI40m+HE1awcmKqCJzjhwAJIMWeop\nREVs+bxZ5pR9OgoZTrGaVEoVP80Gqq1UhdAHJv2flA8LKqIXGKyj/ZzgFEpOzPHU\nif7Qjpodq2LDWhg+wDE0QVAPCdxgRMb7Glokezhpm5rgfwSmLPLkcrAxj4Y15xX/\neAgmlhYFROYdDIMCYxcvP4Cj8hVqng1p/FdvCbzIt5waxrNfVMrqiPXsQNcrUZKE\nKGpzANWGB6yOiXZg1NnHDusrnhUOccsI9zohfjXtnT6SplS5AY0EYZ6sfAEMAM0V\nk/zoNopspU9hXCFtThQkoaMJRhmQPnU27ZZgKrnN9z6k09Utm3JS6x8IQFbNYlWT\nG60Q1FGyH4qNa/2eQj7NJJQAgcnKIXcroi5/64x/QHHtp4Mx/WRqyEKZQ/uzNIQM\nU/CsQ3l5438jAvRBFo18UESXDprxNJX0uwtXV9OyglJYYVs2kNBwdXVVyGwQI3cz\nP/LE26qj0xzxBZv75qp9qkRNdc6bKOvIOfKaM76mXZE5+PWz6XV+A2poAYfJJLkn\nkmF3s7jBPUCSQv4Dj9qBIQiWxzlt/dfXiuuA1qnFrjBdW75PV5l7Ae2cv5UUlGi3\nUjvk2PoYlmatRmyXGBCoB6m/RK19c7FiBGCG/Ij+XackqcJo43k8ZR2JCDRJd+88\ntBY73Zspb3GmMxCLyLFVAiRCQuLC0IRSaYncfsqXABVJN97ElR7dNZD4awk8MaYD\n95xZgmXJmfhaxbaXPMCp9TGSILlDHkhd9vaF9u1YBx8IQVCKP7Dcj9h4tpiEaQAR\nAQABiQG8BBgBCAAmFiEEjzNqcFB0Yj9/wyc5RazhTj67m7oFAmGerHwCGwwFCQPC\nZwAACgkQRazhTj67m7qG/wv/WJd3SMb61dX6rYJ0C+uyXlCZyM+Ama5EKzuS2+jW\n2n1eTj7Mpl+bERgyI0MJnlDq4oEGrc14CnD286thSnOaSJ4TQ8qqhSQCCyzEcHW/\nH9s+jrq00bhfdaHDDzLmtDi7zGi1y8/wS65W7XGT2WqQXFQvm/LJ1UKNmzZYYS8P\nWZH8H5NHDEC0ks1HQ71mW1heGCzOYdCtq0fpc+nxWJ9AptI5Zz+Pzin4/XJ2cd5i\np+6ZAD6S2oC8pqhLvMoYsSWS82uPUoHYQjL6wwPjmSNKmg0+uTG6TuVCLgNwFum7\nnKFo5OfWw32qgfDCKySTJ3GZtXXAqg39zKIuzKpNgJkrosncK+LP3AxAY1VQocjN\n3L04tFuBxjtpw6TY/V/ZvfAT4AAbJVVHPCWR5dfDaVWWe3CeTDU1mCoZu5Cxlztt\nnH6JX0C5rr3P3oaMb/tYcGvIs5LpmpnFnZ+j0J9GPbfvv4STBoUp7npk+WKgmgmo\nEC9uT1kaZiXrUN6yEcniORsH\n=xlkd\n-----END PGP PUBLIC KEY BLOCK-----\n",
            "pulp_created": "2021-12-10T20:39:13.670813Z",
            "pulp_href": "/pulp/api/v3/signing-services/fd81265f-8e6c-491e-8488-13ce1b1ce7db/",
            "script": "/home/vagrant/devel/pulp_container/bash-script.sh"
        }
    ]
}


(pulp) [vagrant@pulp3-source-fedora34 pulp_container]$ http https://pulp3-source-fedora34.fluffy.example.com/pulp/api/v3/repositories/container/container-push/3b279a32-b313-44bf-ad41-4359a92cae24/sign/ manifest_signing_service=/pulp/api/v3/signing-services/fd81265f-8e6c-491e-8488-13ce1b1ce7db/
HTTP/1.1 202 Accepted
Access-Control-Expose-Headers: Correlation-ID
Allow: POST, OPTIONS
Connection: keep-alive
Content-Length: 67
Content-Type: application/json
Correlation-ID: f397ba767a9649b68fee8fe90826e1e7
Date: Fri, 10 Dec 2021 20:39:56 GMT
Referrer-Policy: same-origin
Server: nginx/1.20.1
Vary: Accept, Cookie
X-Content-Type-Options: nosniff
X-Frame-Options: DENY

{
    "task": "/pulp/api/v3/tasks/2d6eb9b7-f5aa-40b5-be1c-99c40805d049/"
}


(pulp) [vagrant@pulp3-source-fedora34 pulp_container]$ http GET https://pulp3-source-fedora34.fluffy.example.com/pulp/api/v3/tasks/2d6eb9b7-f5aa-40b5-be1c-99c40805d049/
HTTP/1.1 200 OK
Access-Control-Expose-Headers: Correlation-ID
Allow: GET, PATCH, DELETE, HEAD, OPTIONS
Connection: keep-alive
Content-Length: 722
Content-Type: application/json
Correlation-ID: b4ee9ec461be46fb9383164b8f5f140b
Date: Fri, 10 Dec 2021 20:40:06 GMT
Referrer-Policy: same-origin
Server: nginx/1.20.1
Strict-Transport-Security: max-age=15768000
Vary: Accept, Cookie
X-Content-Type-Options: nosniff
X-Frame-Options: DENY

{
    "child_tasks": [],
    "created_resources": [
        "/pulp/api/v3/repositories/container/container-push/3b279a32-b313-44bf-ad41-4359a92cae24/versions/10/"
    ],
    "error": null,
    "finished_at": "2021-12-10T20:39:57.016883Z",
    "logging_cid": "f397ba767a9649b68fee8fe90826e1e7",
    "name": "pulp_container.app.tasks.sign.sign",
    "parent_task": null,
    "progress_reports": [],
    "pulp_created": "2021-12-10T20:39:56.741507Z",
    "pulp_href": "/pulp/api/v3/tasks/2d6eb9b7-f5aa-40b5-be1c-99c40805d049/",
    "reserved_resources_record": [
        "/pulp/api/v3/repositories/container/container-push/3b279a32-b313-44bf-ad41-4359a92cae24/"
    ],
    "started_at": "2021-12-10T20:39:56.780215Z",
    "state": "completed",
    "task_group": null,
    "worker": "/pulp/api/v3/workers/eb65c2d9-31b2-47dc-847e-dad0e744c539/"
}


(pulp) [vagrant@pulp3-source-fedora34 pulp_container]$ http GET https://pulp3-source-fedora34.fluffy.example.com/pulp/api/v3/repositories/container/container-push/3b279a32-b313-44bf-ad41-4359a92cae24/versions/10/
HTTP/1.1 200 OK
Access-Control-Expose-Headers: Correlation-ID
Allow: GET, DELETE, HEAD, OPTIONS
Connection: keep-alive
Content-Length: 1364
Content-Type: application/json
Correlation-ID: 744c771d37324c1c9254ed358b674f68
Date: Fri, 10 Dec 2021 20:40:13 GMT
Referrer-Policy: same-origin
Server: nginx/1.20.1
Strict-Transport-Security: max-age=15768000
Vary: Accept, Cookie
X-Content-Type-Options: nosniff
X-Frame-Options: DENY

{
    "base_version": null,
    "content_summary": {
        "added": {
            "container.signature": {
                "count": 1,
                "href": "/pulp/api/v3/content/container/signatures/?repository_version_added=/pulp/api/v3/repositories/container/container-push/3b279a32-b313-44bf-ad41-4359a92cae24/versions/10/"
            }
        },
        "present": {
            "container.blob": {
                "count": 8,
                "href": "/pulp/api/v3/content/container/blobs/?repository_version=/pulp/api/v3/repositories/container/container-push/3b279a32-b313-44bf-ad41-4359a92cae24/versions/10/"
            },
            "container.manifest": {
                "count": 1,
                "href": "/pulp/api/v3/content/container/manifests/?repository_version=/pulp/api/v3/repositories/container/container-push/3b279a32-b313-44bf-ad41-4359a92cae24/versions/10/"
            },
            "container.signature": {
                "count": 1,
                "href": "/pulp/api/v3/content/container/signatures/?repository_version=/pulp/api/v3/repositories/container/container-push/3b279a32-b313-44bf-ad41-4359a92cae24/versions/10/"
            },
            "container.tag": {
                "count": 1,
                "href": "/pulp/api/v3/content/container/tags/?repository_version=/pulp/api/v3/repositories/container/container-push/3b279a32-b313-44bf-ad41-4359a92cae24/versions/10/"
            }
        },
        "removed": {}
    },
    "number": 10,
    "pulp_created": "2021-12-10T20:39:56.942014Z",
    "pulp_href": "/pulp/api/v3/repositories/container/container-push/3b279a32-b313-44bf-ad41-4359a92cae24/versions/10/",
    "repository": "/pulp/api/v3/repositories/container/container-push/3b279a32-b313-44bf-ad41-4359a92cae24/"
}


(pulp) [vagrant@pulp3-source-fedora34 pulp_container]$ http GET https://pulp3-source-fedora34.fluffy.example.com/pulp/api/v3/content/container/signatures/?repository_version=/pulp/api/v3/repositories/container/container-push/3b279a32-b313-44bf-ad41-4359a92cae24/versions/10/
HTTP/1.1 200 OK
Access-Control-Expose-Headers: Correlation-ID
Allow: GET, HEAD, OPTIONS
Connection: keep-alive
Content-Length: 585
Content-Type: application/json
Correlation-ID: 98291c58e78c4db3a91be6d2f822ed8e
Date: Fri, 10 Dec 2021 20:40:20 GMT
Referrer-Policy: same-origin
Server: nginx/1.20.1
Strict-Transport-Security: max-age=15768000
Vary: Accept, Cookie
X-Content-Type-Options: nosniff
X-Frame-Options: DENY

{
    "count": 1,
    "next": null,
    "previous": null,
    "results": [
        {
            "creator": "atomic 5.16.2-dev",
            "digest": "sha256:2d916bd0c131e9da11d09a8490a4529cf8fd5b3063b093a2ce115c45d8564c4a",
            "key_id": "45ACE14E3EBB9BBA",
            "name": "sha256:de0b3821d652af121ad384b0198dc1c6926f77531d6c250cecff3c42d29c95ce@2d916bd0c131e9da11d09a8490a4529c",
            "pulp_created": "2021-12-10T20:39:56.933134Z",
            "pulp_href": "/pulp/api/v3/content/container/signatures/365af055-320b-4e19-8cd9-7a3fcaa620d2/",
            "signed_manifest": "/pulp/api/v3/content/container/manifests/51caa6c9-5c93-4843-9488-c01de3effdf3/",
            "timestamp": 1639168796,
            "type": "atomic"
        }
    ]
}

(pulp) [vagrant@pulp3-source-fedora34 pulp_container]$ cat bash-script.sh 
#!/usr/bin/env bash

#set -x
MANIFEST_PATH1=$1
FINGEPRINT1="$PULP_SIGNING_KEY_FINGERPRINT"
REFERENCE1="$REFERENCE"
SIG_PATH1="$SIG_PATH"

# Create container signature
skopeo standalone-sign $MANIFEST_PATH1 $REFERENCE1 $FINGEPRINT1 --output $SIG_PATH1
# Check the exit status
STATUS=$?
if [ $STATUS -eq 0 ]; then
  echo {\"signature\": \"$SIG_PATH1\"}
else
  exit $STATUS
fi

(pulp) [vagrant@pulp3-source-fedora34 pulp_container]$ 

```